### PR TITLE
LBP1 ratings display + `/slots/highestRated`

### LINKED ISSUE #182: Highest Rated (`/lbp/slots/highestRated`)

### PR DESCRIPTION
**Pull request by Beyley**
Closes #182

### CODE REVIEW

**Comment by Beyley** [comment]
Tested working on LBP1, PSP is still broken, but that can come in a later PR

**Comment by Beyley** [comment]
To clarify, highest rated _semi-works_ on PSP now (seems the game only displays levels returned from the endpoint if it has a rating, and right now i cant get the game to consistently accept a rating, so its still technically broken)

**Comment by Beyley** [comment]
Closes #182

File: Refresh.GameServer/Database/GameDatabaseContext.Relations.cs [hunk 1]
```diff
@@ -161,21 +146,13 @@ public bool DequeueLevel(GameLevel level, GameUser user)
...
     }
 
-    public void ClearQueue(GameUser user)
-    {
-        this._realm.Write(() =>
-        {
-            this._realm.RemoveRange(user.QueueLevelRelations);
-        });
-    }
+    public void ClearQueue(GameUser user) => this._realm.Write(() => this._realm.RemoveRange(user.QueueLevelRelations));
```
Comment (jvyden) [inline Refresh.GameServer/Database/GameDatabaseContext.Relations.cs:154]: I don't mind it elsewhere, but here the one-lines are hard to read

File: Refresh.GameServer/Database/GameDatabaseContext.Relations.cs [hunk 2]
```diff
@@ -184,13 +161,7 @@ public void ClearQueue(GameUser user)
...
 
     [Pure]
-    public RatingType GetRatingByUser(GameLevel level, GameUser user)
-    {
-        RateLevelRelation? rating = this.GetRateRelationByUser(level, user);
-        if(rating == null) return RatingType.Neutral;
-            
-        return rating.RatingType;
-    }
+    public RatingType? GetRatingByUser(GameLevel level, GameUser user) => this.GetRateRelationByUser(level, user)?.RatingType;
```
Comment (jvyden) [inline Refresh.GameServer/Database/GameDatabaseContext.Relations.cs:164]: I prefer the old method. It's clear what's actually happening when the rating is `null` to the reader. You might know that the default `RatingType` is `Neutral`, but someone else might not.

File: Refresh.GameServer/Endpoints/Game/DataTypes/Response/GameLevelResponse.cs [hunk 3]
```diff
@@ -44,9 +44,10 @@ public class GameLevelResponse : IDataConvertableFrom<GameLevelResponse, GameLev
     [XmlElement("playCount")] public required int TotalPlayCount { get; set; }
     [XmlElement("uniquePlayCount")] public required int UniquePlayCount { get; set; }
 
-    [XmlElement("yourDPadRating")] public int YourRating { get; set; }
+    [XmlElement("yourDPadRating")] public int YourDPadRating { get; set; }
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/DataTypes/Response/GameLevelResponse.cs:47]: Why the change? I think `YourRating` was clear enough, especially given that LBP2 is our primary target

File: Refresh.GameServer/Endpoints/Game/DataTypes/Response/GameLevelResponse.cs [hunk 4]
```diff
@@ -44,9 +44,10 @@ public class GameLevelResponse : IDataConvertableFrom<GameLevelResponse, GameLev
...
     [XmlElement("thumbsup")] public required int YayCount { get; set; }
     [XmlElement("thumbsdown")] public required int BooCount { get; set; }
+    [XmlElement("yourRating")] public int YourRatingLBP1 { get; set; }
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/DataTypes/Response/GameLevelResponse.cs:50]: `YourStarRating` would work better here

File: Refresh.GameServer/Endpoints/Game/DataTypes/Response/GameLevelResponse.cs [hunk 5]
```diff
@@ -63,6 +64,7 @@ public class GameLevelResponse : IDataConvertableFrom<GameLevelResponse, GameLev
     [XmlElement("shareable")] public int IsCopyable { get; set; }
     [XmlElement("backgroundGUID")] public string? BackgroundGuid { get; set; }
     [XmlElement("links")] public string? Links { get; set; }
+    [XmlElement("averageRating")] public double AverageRating { get; set; }
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/DataTypes/Response/GameLevelResponse.cs:67]: `AverageStarRating` if you agree with my above change

File: Refresh.GameServer/Endpoints/Game/Levels/LevelEndpoints.cs [hunk 6]
```diff
@@ -29,7 +29,7 @@ public class LevelEndpoints : EndpointGroup
         (int skip, int count) = context.GetPageData();
 
         DatabaseList<GameLevel>? levels = categoryService.Categories
-            .FirstOrDefault(c => c.GameRoute.StartsWith(route))?
+            .FirstOrDefault(c => c.GameRoutes.Any(x => x.StartsWith(route)))?
```
Comment (jvyden) [inline Refresh.GameServer/Endpoints/Game/Levels/LevelEndpoints.cs:32]: We usually use the first letter of the name instead of `x`, so `r` for `route` in this case.

File: Refresh.GameServer/Types/Levels/Categories/CategoryService.cs [hunk 7]
```diff
@@ -48,7 +48,7 @@ public class CategoryService : EndpointService
             IconHash = "g820607",
             FontAwesomeIcon = "heart",
         },
-        new LevelCategory("mostLiked", "thumbs", false, nameof(GameDatabaseContext.GetHighestRatedLevels))
+        new LevelCategory("mostLiked", new[] { "thumbs", "highestRated" }, false, nameof(GameDatabaseContext.GetHighestRatedLevels))
```
Comment (jvyden) [inline Refresh.GameServer/Types/Levels/Categories/CategoryService.cs:51]: Aliases would probably fit better down below in the initializer, not as a property of the constructor.

File: Refresh.GameServer/Types/Levels/GameLevel.cs [hunk 8]
```diff
@@ -111,4 +112,20 @@ public int SequentialId
     }
 
     public GameUser? Publisher { get; set; }
+
+    /// <summary>
+    /// Calculates the average rating of a level, based on the ratings it has, returns a number 1-5
+    /// </summary>
+    /// <returns>The level's average ratings</returns>
+    public double CalculateAverageRating()
```
Comment (jvyden) [inline Refresh.GameServer/Types/Levels/GameLevel.cs:120]:
```suggestion
    /// <summary>
    /// Calculates the average rating of a level based on the ratings it has.
    /// </summary>
    /// <returns>Returns a double between 1 and 5, indicating the level's average ratings.</returns>
    public double CalculateAverageRating()
```

File: Refresh.GameServer/Types/Reviews/RatingType.cs [hunk 9]
```diff
@@ -5,4 +5,22 @@ public enum RatingType : sbyte
...
     Boo = -1,
+}
+
+public static class RatingTypeExtensions
+{
+    public static int ToDPad(this RatingType type)
+    {
+        return (int)type;
+    }
+
```
Comment (jvyden) [inline Refresh.GameServer/Types/Reviews/RatingType.cs:16]: Not sure what the point of this method is?

**Review by jvyden — CHANGES_REQUESTED** [review]

File: Refresh.GameServer/Database/GameDatabaseContext.Relations.cs (hunk 2 above)
Comment (Beyley) [reply Refresh.GameServer/Database/GameDatabaseContext.Relations.cs:164]: PSP/LBP1 ratings need to know the difference between "no rating" and "neutral rating", would i just call into `GetRateRelationByUser` directly for PSP/LBP1?

File: Refresh.GameServer/Types/Reviews/RatingType.cs (hunk 9 above)
Comment (Beyley) [reply Refresh.GameServer/Types/Reviews/RatingType.cs:16]: Mostly visual clarity when reading code, its not needed at all

Comment (jvyden) [reply Refresh.GameServer/Types/Reviews/RatingType.cs:16]: Feel free to keep it in that case

File: Refresh.GameServer/Database/GameDatabaseContext.Relations.cs (hunk 2 above)
Comment (jvyden) [reply Refresh.GameServer/Database/GameDatabaseContext.Relations.cs:164]: Oh, I misunderstood the purpose of this refactor. Yeah, just do that.

File: Refresh.GameServer/Endpoints/Game/DataTypes/Response/GameLevelResponse.cs (hunk 3 above)
Comment (Beyley) [reply Refresh.GameServer/Endpoints/Game/DataTypes/Response/GameLevelResponse.cs:47]: With the suggestion of `YourStarRating`, i agree its unnessesary to change now, will revert

File: Refresh.GameServer/Types/Levels/Categories/CategoryService.cs (hunk 7 above)
Comment (Beyley) [reply Refresh.GameServer/Types/Levels/Categories/CategoryService.cs:51]: I did it this way to keep the code simpler at the slots endpoint, as instead of checking for "route or one any of the aliases" it checks for "any of the routes"

Comment (jvyden) [reply Refresh.GameServer/Types/Levels/Categories/CategoryService.cs:51]: Hmm, that definitely makes it more annoying to read the constructor but if that's the purpose then I guess it's okay.

File: Refresh.GameServer/Database/GameDatabaseContext.Relations.cs (hunk 2 above)
Comment (Beyley) [reply Refresh.GameServer/Database/GameDatabaseContext.Relations.cs:164]: hmm, seems the only usage of `GetRatingByUser` is [here](https://github.com/LittleBigRefresh/Refresh/blob/abe8b392b748b76968a4de0bc79109032904c89d/Refresh.GameServer/Endpoints/Game/DataTypes/Response/GameLevelResponse.cs#L129) so using `GetRateRelationByUser` directly for LBP1/PSP would either double the database calls made or mean GetRatingByUser goes completely unused in the codebase

We should also probably encourage any potential future users of `GetRatingByUser` to be aware of the semantics of "neutral vs missing", since there *is* a difference now, perhaps an xmldoc explaining what null should mean in use, with my refactored implementation would be clear enough?

Comment (jvyden) [reply Refresh.GameServer/Database/GameDatabaseContext.Relations.cs:164]: I see. I think the nullable marker is enough to convey "the rating is missing", but it couldn't hurt to add xmldoc.

**Review by jvyden — APPROVED** [review]
